app/src/main/java/io/pslab/activity/OscilloscopeActivity.java

```diff
@@ -460,6 +460,10 @@ public void run() {
                             audioJack.release();
                             audioJack = null;
                         }
+
+                        if (!(((isCH1Selected || isCH2Selected || isCH3Selected || isMICSelected) && scienceLab.isConnected()) || isInBuiltMicSelected)) {
+                            mChart.clearValues();
```

Comment:
I get the following exception if I test the changes on my Motorola Moto X Play test device:

`FATAL EXCEPTION: Thread-4
Process: io.pslab, PID: 18636
android.view.ViewRootImpl$CalledFromWrongThreadException: Only the original thread that created a view hierarchy can touch its views.
	at android.view.ViewRootImpl.checkThread(ViewRootImpl.java:6962)
	at android.view.ViewRootImpl.invalidateChildInParent(ViewRootImpl.java:1101)
	at android.view.ViewGroup.invalidateChild(ViewGroup.java:5209)
	at android.view.View.invalidateInternal(View.java:13659)
	at android.view.View.invalidate(View.java:13623)
	at android.view.View.invalidate(View.java:13607)
	at com.github.mikephil.charting.charts.Chart.clearValues(Chart.java:324)
	at io.pslab.activity.OscilloscopeActivity$1.run(OscilloscopeActivity.java:465)
	at java.lang.Thread.run(Thread.java:761)
`

Reply:
I changed line 465 to

` mChart.post(() -> mChart.clearValues());`

to run the clear command on the UI thread. The app does not crash anymore after that, but it becomes unresponsive. I think the reason is that the chart is cleared over and over again in an endless loop in the Runnable which runs in the extra thread.

I think it needs to be ensured that the chart is only cleared if something has been added to it before. I have seen that the Chart has an `isEmpty()` method which I think could be used for this check.

Reply:
@marcnause Thanks so much for testing. I understood the error, I am not sure why it comes up on some devices because on the device I tested, the solution worked fine. In any case, I guess I have a better solution in mind which would eliminate all this hassle of the _clearValues()_ function being called again and again from a thread which isn't the UI one. We could perhaps migrate refreshing of the graph to the fragment for the workaround. I'll make a commit and update this PR :))

Reply:
I think it makes a difference if the PSLab device is connected or not. Looking forward to your PR!

Reply:
@marcnause Yes that is evident. I'll soon be making a commit. I'll take care of that.